lib/nerves_hub_link_common/journal.ex

```diff
@@ -0,0 +1,101 @@
+defmodule NervesHubLinkCommon.Journal do
+  @moduledoc """
+  Simple journaling structure backed by a file on the filesystem
+
+  Stores data in chunks in the following format:
+
+    <<length::32, hash::binary-size(32)-unit(8), data::binary-size(length)-unit(8)>>
+
+  as chunks are streamed with `save_chunk/2` the data is updated both on disk and
+  in the structure. This can be used to rehydrate stateful events after a reboot, such as
+  a firmware update for example.
+
+  When opening an existing journal (done automatically if the journal exists),
+  the structure will validate all the chunks on disk, stopping on either
+
+    * the first chunk to fail a hash check
+    * the end of the file
+
+  In either case, the journal is valid to use at this point
+  """
+
+  defstruct [:fd, :content_length, :chunks]
+
+  @type t :: %__MODULE__{
+          fd: :file.fd(),
+          content_length: non_neg_integer(),
+          chunks: [binary()]
```

Comment:
I'm a little worried about firmware images that exceed available memory. I haven't read the code completely, but keeping a chunk list makes me think they're being cached.

How about a setup where the firmware download process starts by pulling chunks from the journal and feeding them to fwup. When it runs out of journal chunks, it makes an http request for the rest?

Reply:
Yeah i think something along those lines will be required as well. I haven't thought of exactly how yet, but i like your idea.